File: 08-EstContenedoras/02-Citas/04-Javascript/Citas.js
/*
   Crear un programa para garantizar que los pacientes sean atendidos
   de manera equitativa y oportuna. El orden de llegada debe
   equilibrarse con la prioridad médica de cada caso

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Diciembre 2025
   Licencia: GNU GPL v3
*/

const NIVELES = 3;

const NIVEL_RIESGO      = 1;
const NIVEL_URGENCIA    = 2;
const NIVEL_PRIORITARIO = 3;

function main() {
  let citas = ingresarColeccion(ingresarCita);

  let citasPorNivel = ordenarColeccion(citas, 
    compararCampoNivel, false);

  let cantidadRiesgo      = contarSegunCriterio(citas, 
    tenerNivelRiesgo, NIVEL_RIESGO);

  let cantidadUrgencia    = contarSegunCriterio(citas, 
    tenerNivelRiesgo, NIVEL_URGENCIA);

  let cantidadPrioritario = contarSegunCriterio(citas, 
    tenerNivelRiesgo, NIVEL_PRIORITARIO);

  let reporteCitas = generarReporteCitas(citasPorNivel,
                                         cantidadRiesgo,
                                         cantidadUrgencia,
                                         cantidadPrioritario);

  mostrarMensaje(reporteCitas);
}

function ingresarCita() {
  alert("Ingrese los datos de la cita:");

  let nombre = prompt("Ingrese el nombre paciente:");
  let nivel = Number(prompt("NIVEL DE URGENCIA\n"+
		"  1: Riesgo\n"+
		"  2: Urgencia\n"+
		"  3: Prioritario\n"+
		"Ingrese tipo de persona: ")) || 0;

  return { nombre, nivel };
}

function compararCampoNivel(cita1, cita2) {
  return cita1.nivel - cita2.nivel;
}

function contarSegunCriterio(coleccion, aplicarCriterio, valorCriterio) {
  return coleccion.filter((elemento) => aplicarCriterio(elemento, valorCriterio) ).length;
}

function tenerNivelRiesgo(cita, nivelRiesgo) {
	return cita.nivel == nivelRiesgo
}

function generarReporteCitas(citasPorNivel,
                             cantidadRiesgo,
                             cantidadUrgencia,
                             cantidadPrioritario) {
  listadoPorNivel = convertirColeccionCadena(
      "LISTADO POR NIVEL DE URGENCIA", 
      citasPorNivel, 
      convertirCitaCadena);

  return `${listadoPorNivel}\n\n` +
    `Nivel 1 (Riesgo)     : ${cantidadRiesgo}\n` +
    `Nivel 2 (Urdencia)   : ${cantidadUrgencia}\n` + 
    `Nivel 3 (Prioritario): ${cantidadPrioritario}\n`;
}

function convertirCitaCadena(cita) {
  return `${cita.nivel} : ${cita.nombre}\n`;
}

File: 08-EstContenedoras/02-Citas/02-Go/Citas.go
/*
   Crear un programa para garantizar que los pacientes sean atendidos
   de manera equitativa y oportuna. El orden de llegada debe
   equilibrarse con la prioridad médica de cada caso

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Diciembre 2025
   Licencia: GNU GPL v3
*/

package main

import (
	"citas/modulo"
	"fmt"
)

const (
	NIVELES = 3

	NIVEL_RIESGO      = 1
	NIVEL_URGENCIA    = 2
	NIVEL_PRIORITARIO = 3
)

type Cita struct {
	Nombre string
	Nivel  int
}

func main() {
	citas := modulo.IngresarColeccion(ingresarCita)

	citasPorNivel := modulo.OrdenarColeccion(
		citas, compararCampoNivel, false)

	cantidadRiesgo := contarSegunCriterio(
		citas, tenerNivelRiesgo, NIVEL_RIESGO)

	cantidadUrgencia := contarSegunCriterio(
		citas, tenerNivelRiesgo, NIVEL_URGENCIA)

	cantidadPrioritario := contarSegunCriterio(
		citas, tenerNivelRiesgo, NIVEL_PRIORITARIO)

	reporteCitas := generarReporteCitas(
		citasPorNivel,
		cantidadRiesgo,
		cantidadUrgencia,
		cantidadPrioritario)

	modulo.MostrarMensaje(reporteCitas)
}

func ingresarCita() Cita {
	modulo.MostrarMensaje("\nIngrese los datos de la cita:\n")
	nombre := modulo.IngresarTexto("\tIngrese el nombre paciente: ")
	nivel := modulo.IngresarOpcion("\tNIVEL DE URGENCIA\n"+
		"\t\t1: Riesgo\n"+
		"\t\t2: Urgencia\n"+
		"\t\t3: Prioritario\n"+
		"\tIngrese tipo de persona: ",
		NIVELES)

	return Cita{Nombre: nombre, Nivel: nivel}
}

func compararCampoNivel(cita1 Cita, cita2 Cita) bool {
	return cita1.Nivel > cita2.Nivel
}

func contarSegunCriterio[Tipo1 any, Tipo2 any](coleccion []Tipo1, aplicarCriterio func(Tipo1, Tipo2) bool, valorCriterio Tipo2) int {
	cantidad := 0

	for _, elemento := range coleccion {
		if aplicarCriterio(elemento, valorCriterio) {
			cantidad += 1
		}
	}

	return cantidad
}

func tenerNivelRiesgo(cita Cita, nivelRiesgo int) bool {
	return cita.Nivel == nivelRiesgo
}

func generarReporteCitas(
	citasPorNivel []Cita,
	cantidadRiesgo int,
	cantidadUrgencia int,
	cantidadPrioritario int) string {
	listadoPorNivel := modulo.ConvertirColeccionCadena(
		"LISTADO POR NIVEL DE URGENCIA",
		citasPorNivel,
		convertirCitaCadena)

	return fmt.Sprintf("%s\n\n"+
		"Nivel 1 (Riesgo)     : %d\n"+
		"Nivel 2 (Urdencia)   : %d\n"+
		"Nivel 3 (Prioritario): %d\n",
		listadoPorNivel,
		cantidadRiesgo,
		cantidadUrgencia,
		cantidadPrioritario)
}

func convertirCitaCadena(cita Cita) string {
	return fmt.Sprintf("%d : %s\n", cita.Nivel, cita.Nombre)
}
